src/nvim/shada.c

```diff
@@ -2215,8 +2222,23 @@ static inline ShaDaWriteResult shada_read_when_writing(ShaDaReadDef *const sd_re
               }
               shada_free_shada_entry(&wms_entry->data);
             }
+          } else {
+            FOR_ALL_BUFFERS(buf) {
+              if (buf->b_ffname != NULL
+                  && FNAMECMP(entry.data.filemark.fname, buf->b_ffname) == 0) {
+                fmark_T fm;
+                mark_get(buf, curwin, &fm, kMarkBufLocal, (int)entry.data.filemark.name);
+                if (fm.timestamp >= entry.timestamp) {
+                  set_wms = false;
+                  break;
+                }
+              }
+            }
+            shada_free_shada_entry(&entry);
```

Comment:
Sanitizer complains about this, I think you need to let this entry through and free it only when discarding it. Try what I suggested before.